### PR TITLE
[SPARK-22290][core] Avoid creating Hive delegation tokens when not necessary.

### PR DESCRIPTION
**Pull request by vanzin**
Hive delegation tokens are only needed when the Spark driver has no access
to the kerberos TGT. That happens only in two situations:

- when using a proxy user
- when using cluster mode without a keytab

This change modifies the Hive provider so that it only generates delegation
tokens in those situations, and tweaks the YARN AM so that it makes the proper
user visible to the Hive code when running with keytabs, so that the TGT
can be used instead of a delegation token.

The effect of this change is that now it's possible to initialize multiple,
non-concurrent SparkContext instances in the same JVM. Before, the second
invocation would fail to fetch a new Hive delegation token, which then could
make the second (or third or...) application fail once the token expired.
With this change, the TGT will be used to authenticate to the HMS instead.

This change also avoids polluting the current logged in user's credentials
when launching applications. The credentials are copied only when running
applications as a proxy user. This makes it possible to implement SPARK-11035
later, where multiple threads might be launching applications, and each app
should have its own set of credentials.

Tested by verifying HDFS and Hive access in following scenarios:
- client and cluster mode
- client and cluster mode with proxy user
- client and cluster mode with principal / keytab
- long-running cluster app with principal / keytab
- pyspark app that creates (and stops) multiple SparkContext instances
  through its lifetime

### CODE REVIEW

**Comment by jerryshao** [comment]
>The effect of this change is that now it's possible to initialize multiple,
non-concurrent SparkContext instances in the same JVM.

@vanzin , do we support in now? As I remembered it was not supported before.

**Comment by vanzin** [comment]
It's always been supported, as long as they're not running at the same time. The only thing is that it was kinda broken with kerberos.

**Comment by jerryshao** [comment]
I see, thanks for the explanation. I didn't think about such scenario.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala [hunk 1]
```diff
@@ -347,6 +347,10 @@ package object config {
     .timeConf(TimeUnit.MILLISECONDS)
     .createWithDefault(Long.MaxValue)
 
+  private[spark] val KERBEROS_RELOGIN_PERIOD = ConfigBuilder("spark.yarn.kerberos.relogin.period")
+    .timeConf(TimeUnit.SECONDS)
+    .createWithDefaultString("1m")
```
Comment (jerryshao) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala]: I think we should put this into doc. Also is it too frequent to call?

**Comment by jerryshao** [comment]
LGTM, just one minor comment.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala (hunk 1 above)
Comment (vanzin) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala]: The call to `checkTGTAndReloginFromKeytab` is a no-op if the TGT is still valid, so it's ok to call it often.

**Comment by jerryshao** [comment]
LGTM, merging to master.

**Pull request closed without merging**